base/reduce.jl

```diff
@@ -280,29 +280,60 @@ mapreduce_impl(f, op, A::AbstractArrayOrBroadcasted, ifirst::Integer, ilast::Int
 """
     mapreduce(f, op, itrs...; [init])
 
-Apply function `f` to each element(s) in `itrs`, and then reduce the result using the binary
-function `op`. If provided, `init` must be a neutral element for `op` that will be returned
-for empty collections. It is unspecified whether `init` is used for non-empty collections.
-In general, it will be necessary to provide `init` to work with empty collections.
+Apply function `f` to each element(s) in `itrs`, and then repeatedly call the 2 argument
+function `op` with those results or results from previous `op` evaluations until a single value is returned.
+
+If provided, `init` is included exactly once as the left-most argument to `op`
+for non-empty `itrs` and serves as the return value for empty `itrs`. It is
+not transformed by the mapping function `f`. It is generally an error to call `mapreduce`
+with empty collections without specifying an `init` value, but in unambiguous cases an
+identity value for `op` may be returned; see [`Base.reduce_empty`](@ref) for more details.
+
+In contrast with [`mapfoldl`](@ref) and [`mapfoldr`](@ref), the sequence of
+function evaluations and the associativity of the reduction is not specified
+and may vary between different methods and Julia versions.
+For example, `mapreduce(√, +, [1, 4, 9])` may be evaluated as either
+`(√1+√4)+√9` (left-associative) _or_ `√1+(√4+√9)` (right-associative).
+The return value for non-associative `op` functions may vary between
+different methods and between Julia versions. For example, `-` is not
+associative and thus `mapreduce(√, -, [1, 4, 9])` may return either
+`-4.0` or `2.0` depending upon the exact method or version of Julia.
+This is also true of some floating point operations that are typically
+associative, for example `mapreduce(identity, +, [.1, .2, .3])` may return
```

Comment:
Floating-point operations are typically *not* associative.

Maybe something like:

> Because floating-point roundoff errors typically break associativity, even for operations like `+` that are associative in exact arithmetic, this also means that the roundoff characteristics of `mapreduce` are implementation-defined; for example, …

Reply:
Yeah, I meant that to read that `+` itself is typically associative.  I see how I swung and missed, though :)

Reply:
```suggestion
Because floating-point roundoff errors typically break associativity,
even for operations like + that are associative in exact arithmetic,
this also means that the floating-point errors incurred by mapreduce
are implementation-defined; for example `mapreduce(identity, +, [.1, .2, .3])` may return
```